pkg/controller/storagecluster/initialization_reconciler.go

```diff
@@ -0,0 +1,417 @@
+	if err != nil {
+		return err
+	}
+	for _, sc := range scs {
+		existing := storagev1.StorageClass{}
+		err = r.client.Get(context.TODO(), types.NamespacedName{Name: sc.Name, Namespace: sc.Namespace}, &existing)
+
+		switch {
+		case err == nil:
+			if existing.DeletionTimestamp != nil {
+				reqLogger.Info(fmt.Sprintf("Unable to restore init object because %s is marked for deletion", existing.Name))
+				return fmt.Errorf("failed to restore initialization object %s because it is marked for deletion", existing.Name)
+			}
+
+			reqLogger.Info(fmt.Sprintf("Restoring original StorageClass %s", sc.Name))
+			existing.ObjectMeta.OwnerReferences = sc.ObjectMeta.OwnerReferences
+			sc.ObjectMeta = existing.ObjectMeta
+
+			err = r.client.Update(context.TODO(), &sc)
+			if err != nil {
+				return err
+			}
+		case errors.IsNotFound(err):
+			reqLogger.Info(fmt.Sprintf("Creating StorageClass %s", sc.Name))
+			err = r.client.Create(context.TODO(), &sc)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	instance.Status.StorageClassesCreated = true
+
+	return nil
+}
+
+// newStorageClasses returns the StorageClass instances that should be created
+// on first run.
+func (r *ReconcileStorageCluster) newStorageClasses(initData *ocsv1.StorageCluster) ([]storagev1.StorageClass, error) {
+	persistentVolumeReclaimDelete := corev1.PersistentVolumeReclaimDelete
+	ret := []storagev1.StorageClass{
+		storagev1.StorageClass{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: generateNameForCephFilesystemSC(initData),
+			},
+			Provisioner:   fmt.Sprintf("%s.cephfs.csi.ceph.com", initData.Namespace),
+			ReclaimPolicy: &persistentVolumeReclaimDelete,
+			Parameters: map[string]string{
+				"clusterID": initData.Namespace,
+				"fsName":    fmt.Sprintf("%s-cephfilesystem", initData.Name),
+				"csi.storage.k8s.io/provisioner-secret-name":      "rook-csi-cephfs-provisioner",
+				"csi.storage.k8s.io/provisioner-secret-namespace": initData.Namespace,
+				"csi.storage.k8s.io/node-stage-secret-name":       "rook-csi-cephfs-node",
+				"csi.storage.k8s.io/node-stage-secret-namespace":  initData.Namespace,
+			},
+		},
+		storagev1.StorageClass{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: generateNameForCephBlockPoolSC(initData),
+			},
+			Provisioner:   fmt.Sprintf("%s.rbd.csi.ceph.com", initData.Namespace),
+			ReclaimPolicy: &persistentVolumeReclaimDelete,
+			Parameters: map[string]string{
+				"clusterID":                 initData.Namespace,
+				"pool":                      generateNameForCephBlockPool(initData),
+				"imageFeatures":             "layering",
+				"csi.storage.k8s.io/fstype": "ext4",
+				"imageFormat":               "2",
+				"csi.storage.k8s.io/provisioner-secret-name":      "rook-csi-rbd-provisioner",
+				"csi.storage.k8s.io/provisioner-secret-namespace": initData.Namespace,
+				"csi.storage.k8s.io/node-stage-secret-name":       "rook-csi-rbd-node",
+				"csi.storage.k8s.io/node-stage-secret-namespace":  initData.Namespace,
+			},
```

Comment:
@kshlm I ported your secret change here, can you double check this is correct

Reply:
Looks good to me.